### PR TITLE
Add type restricted permutation test and allow multiple event types per dataset

### PR DESCRIPTION
**Pull request by mdml**
Add type-restricted permutation test. Users can pass in multiple mutation and weights file to `compute_exclusivity.py`, which will then combine them into one mutation matrix `A` and one mutation probability matrix `P`. Similarly, they can pass in multiple directories of permuted files to run a type-restricted permutation test (equivalent to the type-restricted test used by MEMCover).

I also modified `process_mutations.py` to take in multiple datasets simultaneously, which it will merge to produce a single mutation file that can be used for the "type oblivious" test. Each dataset now consists of a/multiple MAF file(s) and/or event file(s), where events could be copy number aberrations, gene fusions, etc.

After merging this pull request, the `compute_exclusivity.py` documentation will need to be updated.

### CODE REVIEW

File: compute_exclusivity.py [hunk 1]
```diff
@@ -54,25 +83,46 @@ def run( args ):
...
+            permuted_matrices = [ '{}/{}'.format(permuted_matrix_dir, f) for f in files ]
+            permuted_directory_files.append( permuted_matrices[:args.num_permutations] )
+        assert( len(files) == args.num_permutations for files in permuted_directory_files )
+        
+        permuted_files = zip(*permuted_directory_files)
         print '\tUsing {} permuted matrix files'.format(len(permuted_files))
 
         # Run the permutational
         setToPval, setToRuntime, setToFDR, setToObs = permutational_test( sets, geneToCases, num_patients, permuted_files, args.num_cores, args.verbose )
     else:
```
Comment (matthewreyna) [inline compute_exclusivity.py]: Let's replace `else:` with `elif test == UNWEIGHTED:` and add `else: Raise NotImplementedError('...')`.

**Comment by matthewreyna** [comment]
I computed the type-restricted versions of the weighted exclusivity and permutational tests on triples.  The results were consistent between tests, and they looked reasonable.  I had a `NameError` in `enumerate_sets.py` but fixed it by adding `import sys`.

File: compute_exclusivity.py (hunk 1 above)
Comment (mdml) [inline compute_exclusivity.py]: I think you mean `elif test == UNWEIGHTED or test == WEIGHTED`?

We can make the change, though the argparsing should ensure only `PERMUTATIONAL`, `WEIGHTED`, or `UNWEIGHTED` are passed in.

**Comment by mdml** [comment]
I tested the type-oblivious permutation test on the TCGA Pan12 data with the CoMEt exact test on pairs. 

```
 TCGA_PAN12_DIR=... # fill in with directory
 python process_mutations.py -m $TCGA_PAN12_DIR/blca_cleaned.maf -m $TCGA_PAN12_DIR/brca_cleaned.maf -m $TCGA_PAN12_DIR/coadread_cleaned.maf -m $TCGA_PAN12_DIR/gbm_cleaned.maf -m $TCGA_PAN12_DIR/hnsc_cleaned.maf -m $TCGA_PAN12_DIR/kirc_cleaned.maf -m $TCGA_PAN12_DIR/laml_cleaned.maf -m $TCGA_PAN12_DIR/luad_cleaned.maf -m $TCGA_PAN12_DIR/lusc_cleaned.maf -m $TCGA_PAN12_DIR/ov_cleaned.maf -m $TCGA_PAN12_DIR/ucec_cleaned.maf -ct BLCA BRCA COADREAD GBM HNSC KIRC LAML LUAD LUSC OV UCEC -o tcga-pan12-snvs.json
python compute_exclusivity.py -mf tcga-pan12-snvs.json -o tmp.tsv -f 150 -k 2 Unweighted -m Exact
```

The results look correct and the contingency table numbers match the type-restricted test.

File: compute_exclusivity.py (hunk 1 above)
Comment (matthewreyna) [inline compute_exclusivity.py]: Yes, that's what I meant.

**Comment by matthewreyna** [comment]
I successfully tested the type-oblivious version of the unweighted test on triples.  (The weighted and permutational tests, of course, would require permuting the type-oblivious mutation matrix.)

LGTM.

**Pull request merged**